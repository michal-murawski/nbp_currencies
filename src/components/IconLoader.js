import React from 'react';
import { CircularProgress } from 'material-ui/Progress';

const IconLoader = () => <CircularProgress size={15} thickness={1} />;

export default IconLoader;
