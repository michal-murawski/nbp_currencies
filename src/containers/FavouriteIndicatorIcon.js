import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Favorite from 'material-ui-icons/Favorite';
import NotFavorite from 'material-ui-icons/FavoriteBorder';
import IconButton from 'components/IconButton';
import IconLoader from 'components/IconLoader';
import { getValueByPath } from 'utils/dataHelpers';
import {
  favouritesAddRequest as favouritesAddRequestAction,
  favouritesRemoveRequest as favouritesRemoveRequestAction,
} from 'store/favourites/actions';

class FavouriteIndicatorIcon extends React.PureComponent {
  onClickHandler = () => {
    const {
      code,
      favouriteId,
      favouritesAddRequest,
      favouritesRemoveRequest,
    } = this.props;

    if (favouriteId) {
      favouritesRemoveRequest({ id: favouriteId, code });
    } else {
      return favouritesAddRequest(code);
    }
  };

  render() {
    const { favouriteId, favouritesFetching, isSaving } = this.props;
    if (favouritesFetching || isSaving) {
      return <IconLoader />;
    }

    return (
      <IconButton onClick={this.onClickHandler}>
        {favouriteId ? <Favorite /> : <NotFavorite />}
      </IconButton>
    );
  }
}

FavouriteIndicatorIcon.propTypes = {
  favouritesFetching: PropTypes.bool,
  isSaving: PropTypes.bool,
  code: PropTypes.string,
  favouriteId: PropTypes.number,
};

const mapDispatchToProps = dispatch => ({
  favouritesAddRequest: code => dispatch(favouritesAddRequestAction(code)),
  favouritesRemoveRequest: code => dispatch(favouritesRemoveRequestAction(code)),
});

const mapStateToProps = state => ({
  favouritesFetching: getValueByPath(['favourites', 'fetching'], state),
});

export const FavouriteIndicatorIconRaw = FavouriteIndicatorIcon;

export default connect(mapStateToProps, mapDispatchToProps)(FavouriteIndicatorIcon);
