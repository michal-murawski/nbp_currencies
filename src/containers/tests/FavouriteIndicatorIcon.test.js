import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import IconLoader from 'components/IconLoader';
import Favorite from 'material-ui-icons/Favorite';
import NotFavorite from 'material-ui-icons/FavoriteBorder';

import { FavouriteIndicatorIconRaw } from '../FavouriteIndicatorIcon';

describe('containers/FavouriteIndicatorIcon', () => {
  it('should render properly with Favourites Icon', () => {
    const wrapper = shallow(
      <FavouriteIndicatorIconRaw favouriteId={1} code="PLN" />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render NotFavorite icon', () => {
    const wrapper = shallow(<FavouriteIndicatorIconRaw code="PLN" />);

    expect(wrapper.find(NotFavorite)).toHaveLength(1);
    expect(wrapper.find(Favorite)).toHaveLength(0);
  });

  it('should render IconLoader if fetching', () => {
    const wrapper = shallow(
      <FavouriteIndicatorIconRaw code="PLN" favouritesFetching />
    );

    expect(wrapper.find(IconLoader)).toHaveLength(1);
    expect(wrapper.find(NotFavorite)).toHaveLength(0);
    expect(wrapper.find(Favorite)).toHaveLength(0);
  });
});
