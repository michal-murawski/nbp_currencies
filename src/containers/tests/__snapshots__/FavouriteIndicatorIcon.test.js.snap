// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/FavouriteIndicatorIcon should render properly with Favourites Icon 1`] = `
<WithStyles(IconButton)
  onClick={[Function]}
>
  <pure(Favorite) />
</WithStyles(IconButton)>
`;
